File: cpp_drill/Problems/STK04_EvaluateReversePolishNotation.cpp
// REF : https://leetcode.com/problems/evaluate-reverse-polish-notation/

#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <set>
using namespace std;

class Solution {
private:
    stack<int> st;
public:
    int evalRPN(vector<string>& tokens) {
        set<string> opr = {"+", "-", "*", "/"};
        for (auto x : tokens) {
            if (opr.find(x) == opr.end()) {
                st.push(stoi(x));
            }
            else if (x == "+") {
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                st.push(a + b);
            }
            else if (x == "-") {
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                st.push(b - a);
            }
            else if (x == "*") {
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                st.push(b * a);
            }
            else if (x == "/") {
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                st.push(b / a);
            }
        }

        return st.top();
    }
};


int main() {
    return 0;
}

File: py_drill/Problems/STK04_EvaluateReversePolishNotation.py
# REF : https://leetcode.com/problems/evaluate-reverse-polish-notation/
# NOTES :
#       Use stack. while traversing once the operator comes, pop two values,
#       perform operation and push again into stack.

# T: O(n)
# S: O(n)

class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        stack = []
        for x in tokens:
            if x not in ["+", "-", "*", "/"]:
                stack.append(int(x))
            elif x == "+":
                a = stack.pop()
                b = stack.pop()
                stack.append(a + b)
            elif x == "-":
                a = stack.pop()
                b = stack.pop()
                stack.append(b - a)
            elif x == "*":
                a = stack.pop()
                b = stack.pop()
                stack.append(b * a)
            elif x == "/":
                a = stack.pop()
                b = stack.pop()
                stack.append(int(b / a))

        return stack[0]


def main():
    pass

if __name__ == "__main__":
    main()